com.unity.formats.fbx/Tests/FbxTests/ModelExporterTest.cs

```diff
@@ -712,53 +717,72 @@ public class SkinnedMeshTestDataClass
             var expVerts = exportedSkinnedMesh.sharedMesh.vertices;
             Assert.That (expVerts, Is.Not.Null);
 
-            var origBoneWeights = originalSkinnedMesh.sharedMesh.boneWeights;
-            Assert.That (origBoneWeights, Is.Not.Null);
-            Assert.That (origBoneWeights.Length, Is.GreaterThan (0));
-
-            var expBoneWeights = exportedSkinnedMesh.sharedMesh.boneWeights;
-            Assert.That (expBoneWeights, Is.Not.Null);
-            Assert.That (expBoneWeights.Length, Is.GreaterThan (0));
-
             var origBones = originalSkinnedMesh.bones;
             var expBones = exportedSkinnedMesh.bones;
 
-            int comparisonCount = 0;
-            int minVertCount = Mathf.Min (origVerts.Length, expVerts.Length);
-            for(int i = 0; i < minVertCount; i++){
-                for (int j = 0; j < minVertCount; j++) {
-                    if (origVerts [i] == expVerts [j]) {
-                        // compare bone weights
-                        var origBw = origBoneWeights[i];
-                        var expBw = expBoneWeights [j];
+            // test with more than 4 bone weights
+
+            // Get the number of bone weights per vertex
+            var origBonesPerVertex = origMesh.GetBonesPerVertex();
+            var exportedBonesPerVertex = exportedMesh.GetBonesPerVertex();
+
+            // Get all the bone weights, in vertex index order
+            var origBoneWeights1 = origMesh.GetAllBoneWeights();
+            var exportedBoneWeights1 = exportedMesh.GetAllBoneWeights();
+
+            // get the vertex index to index in bone weights mapping (for easier access)
+            var origVertToBoneWeightIndex = new Dictionary<int, int>();
+            var boneWeightIndexMapping = 0;
+            for (int i = 0; i < origVerts.Length; i++)
+            {
+                origVertToBoneWeightIndex.Add(i, boneWeightIndexMapping);
+                boneWeightIndexMapping += origBonesPerVertex[i];
+            }
+
+            var exportedVertToBoneWeightIndex = new Dictionary<int, int>();
+            boneWeightIndexMapping = 0;
+            for (int i = 0; i < expVerts.Length; i++)
+            {
+                exportedVertToBoneWeightIndex.Add(i, boneWeightIndexMapping);
+                boneWeightIndexMapping += exportedBonesPerVertex[i];
+            }
 
-                        var indexMsg = "Bone index {0} doesn't match";
-                        var nameMsg = "bone names don't match";
+            var comparisonCount = 0;
+            var minVertCount = Mathf.Min(origVerts.Length, expVerts.Length);
+            for (int origVertIndex = 0; origVertIndex < minVertCount; origVertIndex++)
+            {
+                for (int expVertIndex = 0; expVertIndex < minVertCount; expVertIndex++)
+                {
+                    if (origVerts[origVertIndex] == expVerts[expVertIndex])
```

Comment:
I think this bit is wrong.
(1) It's possible to have two vertices in the same location in a mesh; this will only find the first one and assign all the bones to it.
(2) It's quadratic time. In a 100k or 1M vertex mesh, that might take a long time.

Why would the indices change between the original mesh and the export mesh?

Reply:
Oh wait I'm confused, this is testing not exporting.

I think we should test that the indices don't change; they shouldn't (some things depend on vertex indexes not changing).

Reply:
I think this is because there the vertex count did not always match before/after export. We were doing the same before as well.